src/passes/OptimizeInstructions.cpp

```diff
@@ -746,6 +736,17 @@ struct OptimizeInstructions
           }
         }
       }
+      if (auto* condition = select->condition->dynCast<Unary>()) {
+        if (condition->op == EqZInt32) {
+          // flip select to remove eqz, if we can reorder
+          EffectAnalyzer ifTrue(getPassOptions(), features, select->ifTrue);
+          EffectAnalyzer ifFalse(getPassOptions(), features, select->ifFalse);
+          if (!ifTrue.invalidates(ifFalse)) {
+            select->condition = condition->value;
+            std::swap(select->ifTrue, select->ifFalse);
+          }
+        }
+      }
```

Comment:
I moved this case below after to the `select->condition->dynCast<Const>()` which not required this changes and could potentially return earlier.

Reply:
Not sure I understand what you mean here?

I don't think it matters much. But before if we had `!0 ? a : b` then we would transform to `0 ? b : a` and then apply the constant rule. With your change we exit early and would be called again and only then do the second part?

Reply:
Hmm, is it real situation when we could got `!0`? I thought this impossible after optimizeBoolean here: https://github.com/WebAssembly/binaryen/pull/2869/files/306cf9e7e3d6d3fbf516ea99bd4ee55fb58b3063..125fee882032b5bb1847dfc495bf1f0a5a12709b#diff-54e02686df3480c98e3f90a24705c00bR716

Reply:
Alright. It seems this approach works better. With latest simple pre-optimize boolean conditions: https://github.com/WebAssembly/binaryen/pull/2869/commits/8e658c64208babfbd1fb3f225b83746ad29bc8d3#diff-99d7f94950a6eb885a6cec4748776c22